### PR TITLE
Allow the user to select exactly which references to include in the analysis

### PR DESCRIPTION
**Pull request by mhagger**
It would often be convenient to select exactly which references should be used as starting points when enumerating objects for analysis. We already have `--branches`, `--tags`, and `--remotes` to cause those types of references to be included. But what about

* Include all references except for those under `refs/notes` (see issue #1).
* Include all references except for the special `refs/stash` reference used for managing stashes.
* Include all references except for remote references and GitHub's special `refs/pulls` references.
* Include all branches and tags but not CI build tags that match `refs/tags/build-*`.

This PR adds four new options that make it possible to select arbitrary subsets of references:

* `--include PREFIX` and `--include-regexp PATTERN`: specify references that _should be_ included in the analysis.
* `--exclude PREFIX` and `--exclude-regexp PATTERN`: specify references that _should not be_ included in the analysis.

I still need to add tests.

### CODE REVIEW

**Comment by mhagger** [comment]
This could use more integration testing, but it does have unit tests and I've tested it manually, so I think it's GTG.

**Comment by mhagger** [comment]
I added an integration test after all. _Wenn schon, denn schon._

File: git/ref_filter_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		expected bool
+	}{
+		{"refs/heads", "refs/heads/master", true},
+		{"refs/heads", "refs/tags/master", false},
+		{"refs/heads", "refs/he", false},
+		{"refs/heads", "refs/headstrong", false},
+		{"refs/heads", "refs/tags/refs/heads", false},
+		{"refs/stash", "refs/stash", true},
+		{"refs/stash", "refs/stashy", false},
+		{"refs/remotes", "refs/remotes/origin/master", true},
```
Comment (mhp) [inline git/ref_filter_test.go:28]: You don't appear to have any tests covering prefixes ending with `/`.  Is it worth adding a couple of tests to assert the documented behaviour?

```
 // * Prefix "refs/foo/" matches "refs/foo/bar" but not "refs/foo" or
 //   "refs/foobar".
```

**Review by mhp — APPROVED** [review]
Overall, looks great.  I've suggested a couple more test cases, to cover the documented behaviour, but that's all.

File: git/ref_filter_test.go (hunk 1 above)
Comment (mhagger) [reply git/ref_filter_test.go:30]: Good point. Done :ballot_box_with_check:

**Pull request merged**